### PR TITLE
[ONNX] Benchmark to store test data along exported model

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110313
* #110477
* #110178
* #108376
* __->__ #111095

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -1250,78 +1257,59 @@ def wrapper(self, *args, **kwargs) -> Any:
...
-    ) -> str:
         # Hack to get model name.
         from torch._functorch import aot_autograd
 
         model_name = aot_autograd.model_name
-        model_path = pathlib.Path(output_directory, onnx_model_folder_name, model_name)
+        self.model_dir = self._generate_onnx_model_directory(
+            output_directory, self._COMPILER_NAME, model_name
+        )
+        self.model_path = str(
```
Comment (justinchuby) [inline benchmarks/dynamo/common.py:1271]:
```suggestion
        self.model_path = os.fspath(
```

nit: consider https://docs.python.org/3/library/os.html#os.fspath

File: benchmarks/dynamo/common.py [hunk 2]
```diff
@@ -1250,78 +1257,59 @@ def wrapper(self, *args, **kwargs) -> Any:
...
+            self.model_dir / f"{model_name}_{self._COMPILER_NAME}.onnx"
+        )
+
+    @classmethod
+    def _generate_onnx_model_directory(
+        cls, output_directory: str, compiler_name: str, model_name: str
+    ) -> str:
```
Comment (justinchuby) [inline benchmarks/dynamo/common.py:1278]:
```suggestion
    ) -> pathlib.Path:
```

maybe this?

File: benchmarks/dynamo/common.py [hunk 3]
```diff
@@ -1429,34 +1383,144 @@ def run_with_iobinding(self, iobinding, outputs):
...
+        torch.int16: np.int16,
+        torch.int32: np.int32,
+        torch.int64: np.longlong,
+        torch.bool: np.bool_,
+    }
+
+    _COMPILER_NAME = "torchscript"
+
+    def __init__(self, output_directory, model, example_inputs, dynamic_shapes: bool):
+        assert not dynamic_shapes, "NYI dynamic shapes for OnnxModelFromTorchScript"
```
Comment (justinchuby) [inline benchmarks/dynamo/common.py:1448]: raise not implemented error?

File: benchmarks/dynamo/common.py [hunk 4]
```diff
@@ -1250,78 +1257,59 @@ def wrapper(self, *args, **kwargs) -> Any:
...
-        if isinstance(example_inputs, dict):
-
-            class WrapperModel(torch.nn.Module):
-                def __init__(self, model, keys):
-                    super().__init__()
-                    self.model = model
-                    self.keys = keys
+    @abc.abstractmethod
+    def format_pt_inputs(self, pt_inputs):
+        ...
```
Comment (thiagocrepaldi) [inline benchmarks/dynamo/common.py:1292]: type hint and docstring would help expressing the intent and acceptable input. In fact, docstring would help differentiating `format_pt_inputs` and  `adapt_pt_inputs_to_onnx`

File: benchmarks/dynamo/common.py [hunk 5]
```diff
@@ -1429,34 +1383,144 @@ def run_with_iobinding(self, iobinding, outputs):
...
+        torch.float16: np.float16,
+        torch.float32: np.float32,
+        torch.float64: np.float64,
+        torch.uint8: np.uint8,
+        torch.int8: np.int8,
+        torch.int16: np.int16,
+        torch.int32: np.int32,
+        torch.int64: np.longlong,
+        torch.bool: np.bool_,
+    }
```
Comment (thiagocrepaldi) [inline benchmarks/dynamo/common.py:1443]: this enum already exists at https://github.com/pytorch/pytorch/blob/main/torch/onnx/_internal/fx/type_utils.py#L166

**Review by thiagocrepaldi — APPROVED** [review]
Overall LGTM. Just minor comments such as docstrings and reuse of existing enum

File: benchmarks/dynamo/common.py (hunk 1 above)
Comment (BowenBao) [reply benchmarks/dynamo/common.py:1271]: isn't str() the recommended way of converting Path to string?

Comment (justinchuby) [reply benchmarks/dynamo/common.py:1276]: I asked sometime in the Python channel and was recommended to use os.fspath. If the path can be btypes, we should use os.fsdecode

**Pull request closed without merging**